Disable Serde's default-features

ass3rt opened the pull request:
With this patch, existing users of the `use-serde` feature will no longer be
compiling with `serde/std` enabled, but this allows dependent projects
to import serde and enable `serde/alloc` as required by some no-std targets.

ass3rt commented:
Seems I have some test failures to look into. I'll try to get these worked out tonight.

tcharding commented:
Changing features makes this a breaking change, right? We are in the process of a big breaking release anyways but thought I'd flag it.

tcharding commented on Cargo.toml:
```diff
@@ -41,7 +41,7 @@ core2 = { version = "0.3.0", optional = true, default-features = false }
 base64-compat = { version = "1.0.0", optional = true }
 bitcoinconsensus = { version = "0.19.0-3", optional = true }
-serde = { version = "1", features = [ "derive" ], optional = true }
+serde = { version = "1", default-features = false, optional = true }
```
We need to leave "derive" enabled.

ass3rt replied:
Indeed :facepalm:. Fixed with force-push

ass3rt commented:
> Changing features makes this a breaking change, right?

That is my understanding :+1:

apoelstra reviewed (dismissed):
ACK f18c42c8e51821d1e0d0c0f8fd7358f8e02d0002

Yes, this is a breaking change, but it's fine to bring into the RC.

dr-orlovsky commented on Cargo.toml:
```diff
@@ -41,7 +41,7 @@ core2 = { version = "0.3.0", optional = true, default-features = false }
+serde = { version = "1", default-features = false, features = [ "derive" ], optional = true }
 hashbrown = { version = "0.8", optional = true }
 
 [dev-dependencies]
```
Since we removed `std` feature from `serde`, which breaks tests. Thus, in `dev-depencencies` we must add
```toml
serde = "1"
```

ass3rt replied:
Good catch. I will fix this.

ass3rt replied:
Fixed: 580650f

ass3rt replied:
Dropped this in favor of `serde-std` as discussed below.

dr-orlovsky commented on Cargo.toml:
```diff
@@ -41,7 +41,7 @@ core2 = { version = "0.3.0", optional = true, default-features = false }
 base64-compat = { version = "1.0.0", optional = true }
 bitcoinconsensus = { version = "0.19.0-3", optional = true }
-serde = { version = "1", features = [ "derive" ], optional = true }
+serde = { version = "1", default-features = false, features = [ "derive" ], optional = true }
```
This will result in the use of non-std version of serde even if both `serde` and `std` features are used for `rust-bitcoin` compillation

ass3rt replied:
Yes, I am not aware of a great way around this. In fact, as it stands today, a no-std build will still compile `serde/std`, which seems equally misleading. Attempting to build for a target which does not support std results in compile failures.

A possible compromise would be to re-export the `serde/std` feature, so that a dependent project can enable it if desired.

I'm also open to suggestions, if you have other ideas how to allow serde in pure no-std builds. Or perhaps drop a comment in the parent issue: https://github.com/rust-bitcoin/rust-bitcoin/issues/904

ass3rt replied:
Related to this question, if compiling for `no-std`, wouldn't it be best to ensure all dependencies are also compiling for `no-std`? I thought this was the case until I found out the hard way that Serde is still compiling with std enabled. At best its confusing. At worst it could lead to unintended use of shared system libraries.

apoelstra replied:
Can't we just add `serde/std` to the `std = ["secp256k1/std", "bitcoin_hashes/std", "bech32/std"]` line?

apoelstra replied:
Ah, right, no, because that would pull `serde` in. Never mind. Cargo..

ass3rt replied:
That was my original thought too, but Cargo... :sob:

Surely this isn't the first time a project has needed conditional features in the manifest? I spent some time searching for better workarounds, but so far no luck. I'm all ears if someone has a better way to do this.

apoelstra replied:
We have run into this probably dozens of times. It is a known issue but there is still no fix.

We should probably just add the extra `serde-std` feature.

Kixunil replied:
Yes, this is an old issue and there are exactly two solutions:

* add `serde-std` feature
* bump MSRV to 1.60 - the one to be released next

ass3rt replied:
Added `serde-std` and confirmed builds+tests pass on stable, nightly and 1.29.0.

dr-orlovsky reviewed (changes requested):
NACK: build is broken due to test dependency issues + possible undesirable side-effect of always having non-std serde even for `std` builds

ass3rt commented:
Sorry for the repeated test failures. I promise I am running `./contrib/test.sh` before submitting patches, but the tests are passing on my system. Perhaps I'm missing a step. If anyone has a recommended test workflow before submitting PRs, I will happily accept any advice to improve my testing :)

apoelstra commented:
There are probably some environment variables you need to set to get `test.sh` to do anything. In particular `DO_FEATURE_MATRIX=true`.

Kixunil commented:
The test failures are actually caused by the lack of `serde-std` feature. We really have to use it since I don't believe anyone here wants to bump MSRV to 1.60.

ass3rt commented:
I've reworked this PR to address comments, fix tests, and rebase on-top of master. Sorry for the delay.

Specifically, I added the `serde-std` feature for those that need it, and enabled it for tests.

tcharding commented on Cargo.toml:
```diff
@@ -18,6 +18,7 @@ base64 = [ "base64-compat" ]
 unstable = []
 rand = ["secp256k1/rand-std"]
 use-serde = ["serde", "bitcoin_hashes/serde", "secp256k1/serde"]
+serde-std = ["serde/std"]
```
Perhaps we should turn on std in bitcoin_hashes as well?

```suggestion
serde-std = ["serde/std", "bitcoin_hashes/serde-std"]
```

Kixunil replied:
That'd force-include `bitcoin_hashes`, so no. If we require it anywhere we may need `serde-bitcoin-hashes-std` :(

ass3rt replied:
We could go down a rabbit hole, exhaustively reexporting the `foo/std` sub features. Its not clear to me where to draw the line, so I kept this PR to the minimal change to fix broken builds for `no-std` users of Serde.

I'm open to suggestions if you'd prefer I apply this same logic across some/all other features too

tcharding replied:
Sorry for the dud suggestion, seems my brain cannot easily model this matrix of features across crates.

Kixunil reviewed (commented):
This doesn't look right, the feature is not used anywhere. Did you forget to commit some files?

ass3rt commented:
> This doesn't look right, the feature is not used anywhere. Did you forget to commit some files?

The only place I am aware of that requires 'serde-std' is in tests, which I added. Are you aware of anywhere else that requires it?

Keep in mind, any feature which enables `serde-std` will break `no-std` builds, so IMO it should not be a default feature or sub-feature of any other features.

ass3rt commented:
Ok, I'll back out `serde-std`. It was a useful thought experiment, if nothing else :)

Kixunil commented:
Also to address @dr-orlovsky concern: users needing `serde/std` can and should enable it explicitly when depending on `serde`.

ass3rt commented:
Oops I forgot to run the dep_test. I _did_ have environment variables set to enable that test, but I must've cleared them before my last push :facepalm:.

I forgot to explicitly add `serde` with default features as a dependency for the dep_test. Fixed and force pushed. Tests are passing again. Sorry for all the churn!

Kixunil commented:
Oh, shit. I got confused. For some reason the compatibility issue can't be triggered from normal `cargo test`, only from a dependency and I *also* forgot to set the env var. :man_facepalming:

To clean up my shit, I looked at it properly and found a proper solution. Wanted to make a PR against *your* repo but was unable to do so. Here's my changes: https://github.com/Kixunil/rust-bitcoin/commit/28bb060cae86f07c6588ac9eb572c73c1c25e6ba

Explanation:

I started re-adding `serde-std` but then I realized the affected methods actually re-implement the default behavior (the existing ones already forward!) so I could just delete them. It looked like that was the only change required but then I noticed in case of script we can get perf benefit from overriding. I then realized we don't need `serde-std` but `serde-alloc` and soon after I realized we can depend on it unconditionally because the whole library already depends on `alloc` and I don't think it will stop depending on it any time soon. (Maybe `bitcoin-units` will one day.) Thus I just added `alloc` to the serde features and kept the other changes (even though they are off-topic).

If you want to be super-pedantic and make the history nicer you could split it up into three commits (changing serde features, removing default impls, adding override) but I personally wouldn't block the PR if you don't want to.

ass3rt commented:
@Kixunil good finds! I like that approach. I am afk for several hours today, but I found the github setting to let others push commits to my PR. You should be able to push now. If not, I'll pull your changes in when I get a chance this evening.

Kixunil commented:
Cool, pushed! Do you want to do the squashing?

Kixunil commented:
Oh, shit, MSRV issues. Suggested solution: remove the optimization and `alloc` dependency and reintroduce them after MSRV bump.

Now I finally understand why `serde` didn't do `std = ["alloc"]` but conditions everything using `any(feature = "std", feature = "alloc")`.

Kixunil commented:
Let's just wait for #964

apoelstra commented:
We have #964 now -- can you rebase?

ass3rt commented:
> We have #964 now -- can you rebase?

Awesome. Done :heavy_check_mark:

Kixunil commented:
Could you clean up the git history please? Ideally make it three commits: 1. the topic of this PR, 2. removing unneeded method impls, 2 adding the optimization. But I can accept squashing to a single commit.

ass3rt commented:
@Kixunil good call. Done. Thanks for all your assistance!

Kixunil reviewed (approved):
Perfect, thanks!

ACK 76fcf814745649c20d17e30515ae4f719eebff91

tcharding reviewed (approved):
ACK 76fcf814745649c20d17e30515ae4f719eebff91

apoelstra reviewed (approved):
ACK 76fcf814745649c20d17e30515ae4f719eebff91
